src/aws_lambda_utils/xray_layer.rs

```diff
@@ -0,0 +1,254 @@
+use chrono::{DateTime, Utc};
+use rand::{thread_rng, Rng};
+use serde::Serialize;
+use tokio::{
+    io,
+    net::UdpSocket,
+    spawn,
+    sync::{
+        mpsc::{unbounded_channel, UnboundedSender},
+        OnceCell,
+    },
+    task::JoinHandle,
+};
+use tracing::{error, trace, Id, Level, Subscriber};
+use tracing_subscriber::{layer::Context, registry::SpanRef, Layer};
+
+use super::{tracing_request_id_layer::RequestIds, tracing_span_fields::SpanData};
+
+/// Layer to send spans to the X-ray daemon [https://docs.aws.amazon.com/xray/latest/devguide/xray-daemon.html]
+pub struct XrayLayer<T> {
+    sender: T,
+}
+
+impl<T> XrayLayer<T> {
+    pub fn new_udp_sender() -> XrayLayer<UdpSpanSender> {
+        Self::with_sender(UdpSpanSender::new())
+    }
+    pub fn with_sender<S: SpanSender>(sender: S) -> XrayLayer<S> {
+        XrayLayer { sender }
+    }
+}
+
+pub trait SpanSender {
+    fn submit_span(&self, span: XraySpan);
+}
+
+pub struct UdpSpanSender {
+    queue: UnboundedSender<XraySpan>,
+    _background_task: JoinHandle<()>,
+}
+
+impl UdpSpanSender {
+    fn new() -> Self {
+        let (send, mut recv) = unbounded_channel::<XraySpan>();
+
+        let background_task = spawn(async move {
+            let socket = OnceCell::new();
+            let mut buf = Vec::new();
+            // TODO: Batch requests?
+            loop {
+                while let Some(span) = recv.recv().await {
+                    let sock = socket
+                        .get_or_try_init(|| async {
+                            let addr = std::env::var("AWS_XRAY_DAEMON_ADDRESS")
+                                .unwrap_or_else(|_| "127.0.0.1:2000".to_string());
+                            let s = UdpSocket::bind("0.0.0.0:0").await?;
+                            s.connect(addr).await?;
+                            io::Result::Ok(s)
+                        })
+                        .await;
+                    match sock {
+                        Ok(sock) => {
+                            let header = &b"{\"format\": \"json\", \"version\": 1}\n";
+                            buf.extend(header.iter());
+                            serde_json::to_writer(&mut buf as &mut Vec<_>, &span)
+                                .expect("Xray span to JSON serialization failed");
+                            let document_size = buf.len() - header.len();
+                            if document_size > (64 * 1024) {
+                                error!("segment size exceeded max Xray size of 64kB (was {document_size}). Not sending");
+                            } else {
+                                let res = sock.send(&buf).await;
+                                match res {
+                                    Err(err) => {
+                                        error!(%err, ?span, "failed sending span to Xray")
```

Comment:
Wooot.. What is this magic syntax? :p

Reply:
https://docs.rs/tracing/latest/tracing/index.html#recording-fields

This adds `err` as a field to the log with it's `Display` implementation. And `span` as a Debug representation.

Normally with `error!("failed sending span to Xray")` an implicit `message` field is added to the `tracing` event. So in structured logging terms it will be something like
```json
{"message": "failed sending span to Xray"}
```

while with the fields it will be 
```json
{"message": "failed sending span to Xray", "err": "<Display impl of err>", "span": "<Debug impl of span>"}
```

`tracing` does per-default not have the possibility to have fields that are anything other than the primitive types, so I also added support for the experimental `valuable` crate. That allows us to add any value that implements `Valuable`, which will then be arbitrary structs/enums.

So it's just a way to add context to a log :)